hs/src/Daml/Cucumber.hs

```diff
@@ -79,39 +82,94 @@ runTestSuite (Opts folder mFeatureFile damlFolder) = do
     missingSteps = Set.difference requiredSteps definedSteps
     stepMapping = mconcat $ fmap makeStepMapping files
 
-  case Set.null missingSteps of
-    False -> do
-      putStrLn "Cannot run features, missing steps:"
-      for_ (Set.toList missingSteps) $ putStrLn . prettyPrintStep
+    (_, result) = runState (generateDamlSource definedSteps stepMapping features) (DamlScript mempty mempty)
+    testfile = (damlFolder </> "Generated.daml")
+
+    shouldRunTests = allowMissing || Set.null missingSteps
 
+  case shouldRunTests of
     True -> do
-      let
-        (_, result) = runState (generateDamlSource stepMapping features) (DamlScript mempty mempty)
-        testfile = (damlFolder </> "Generated.daml")
       writeDamlScript testfile result
-      testResults <- runTestLspSession folder testfile $ fmap damlFuncName $ damlFunctions result
+      result <- runTestLspSession folder testfile $ fmap damlFuncName $ damlFunctions result
+      case result of
+        Left error -> do
+          T.putStrLn $ "lsp session failed: " <> error
+          exitFailure
+
+        Right testResults -> do
+          success <- evaluateResults definedSteps testResults features
+          when (not $ Set.null missingSteps) $ do
+            putStrLn "Missing steps:"
+            for_ (Set.toList missingSteps) $ putStrLn . prettyPrintStep
+          when (not success) $ exitFailure
```

Comment:
if we've allowed missing steps, does it still cause a failure at this point?

Reply:
Nope, missing steps does not error if we allowed it, we still report them